vyper/context/README.md

```diff
@@ -0,0 +1,211 @@
+
+### Namespace
+
+[`namespace.py`](namespace.py) contains the `Namespace` object. `Namespace` is a
+`dict` subclass representing the namespace of a contract. It imposes several
+additional restrictions:
+
+* Attempting to replace an existing field raises `NamespaceCollision`
+* Attempting to access a key that does not exist raises `UndeclaredDefinition`
+
+#### Scoping and Namespace as a Context Manager
+
+Each smart contract in Vyper uses a unique namespace, and has three primary scopes:
+
+* **builtin**: core types, environment variables and builtin functions
+* **module**: contract functions, storage variables and user-defined constants
+* **local**: memory variables and other objects declared within a single contract
+function
+
+Additionally, a new scope is entered for each execution of a `for` loop or branch
+of an `if` statement.
+
+Scoping is handled by calling `Namespace.enter_scope` as a
+[context manager](https://docs.python.org/3/reference/datamodel.html#with-statement-context-managers).
+Values added while inside the context are removed when the context is exited.
+
+```python
+with namespace.enter_scope():
+    namespace['foo'] = 42
+
+namespace['foo']  # this raises an UndeclaredDefinition
+```
+
+Prior to beginning type checking, the first scope **must** be initiated using
+`Namespace.enter_builtin_scope`. This ensures that all builtin objects have
+been added, and resets the content of `self` and `log`.
+
+#### Importing the Namespace Object
+
+The `namespace` module replaces itself with an instance of `Namespace` within
+`sys.modules`. provides an easy way for other modules to import the same
+`Namespace` object and ensures that only one copy of the object exists.
+
+To access the `Namespace` object from another module:
+
+```python
+from vyper.context import namespace
+```
+
+Guido van Rossum provides an explanation of why this works in this post from the
+[Python mailing list](https://mail.python.org/pipermail/python-ideas/2012-May/014969.html).
+
+### Type Checking
+
+Type checking is handled bottom-up.
+
+1. The type of each expression is evaluated independently.
+2. Multiple types within the left-hand or right-hand side are compared (e.g. a
+literal array).
+3. The left-hand and right-hand types are compared.
+4. The operation being performed is validated according to the types (e.g. in a
+math operation, the types must be numeric)
+
+In type-checking the following example:
+
+```python
+foo: int128[2] = [-2, 42]
+```
+
+1. The annotation value is validated and determined to be an `int128` array of
+length `2`.
+2. The literal array values are validated on their own. The first value must be
+`int128`, the second could be `int128` or `uint256`.
+3. The right-hand side types are compared. A common type of `int128` is found.
+The array is given a type of `int128[2]`.
+4. The left-hand and right-hand types are compared and found to be matching.
+5. A new castable type `int128[2]` is added to the namespace with the name `foo`.
```

Comment:
This statement made me realize I didn't really understand what a "Castable type" was from above. Perhaps there is a better name? Program type?

Reply:
Yeah, definitely. Naming things is hard.